### PR TITLE
feat: checkbox component 

### PR DESCRIPTION
**Pull request by ngochuyduong**
## Check List
- [x] Your code builds clean without any errors or warnings.
- [x] Check coding convention.
- [x] Updated Storybook components and stories to reflect any changes made to the UI.
- [x] Test covers all edge cases and coverage is greater than or equal to 80%.

## Proof of Completeness
Screenshot 2023-08-06 at 12 29 25 PM

https://user-images.githubusercontent.com/124558477/230290553-97a66282-ea53-4476-a796-15d60c4c53a5.mp4

### CODE REVIEW

File: src/components/Checkbox/Checkbox.tsx [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+      return (
+        <IconContainer
+          testID={'icon-container'}
+          size={size}
+          backgroundColor={checked ? fillColor : unfillColor}
+          disable={disable}
+          disableOpacity={disableOpacity}
+          style={StyleSheet.flatten([{transform: [{scale: bounceValue}]}, iconStyle])}>
+          <InnerIconContainer style={innerIconStyle} size={size} borderColor={fillColor}>
+            {iconComponent ||
```
Comment (TranQuangPhi) [inline src/components/Checkbox/Checkbox.tsx:193]: I think we should change "iconComponent" to "!!iconComponent"

File: src/components/Checkbox/Checkbox.tsx [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        ))
+      )
+    }
+
+    const onHandlePress = useCallback(() => {
+      if (!disableBuiltInState) {
+        setChecked(prev => !prev)
+      }
+      syntheticBounceEffect()
+      onChange && onChange(!checked)
```
Comment (TranQuangPhi) [inline src/components/Checkbox/Checkbox.tsx:219]: Change to onChange?.(!checked). I think this way will be shorter

File: example/ios/Podfile.lock [hunk 3]
```diff
@@ -654,4 +654,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: 5eddd6af0376deacea0a9cf5568699be94aab722
 
-COCOAPODS: 1.11.3
+COCOAPODS: 1.12.0
```
Comment (honghoangsts) [inline example/ios/Podfile.lock:657]: If we upgrade version for cocoapods, Everyone have to upgrade too. it may effect to their current project.

File: src/components/Checkbox/Checkbox.tsx [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+  checkIconImageSource?: ImageSourcePropType
+  /** callback when checkbox is change state */
+  onChange?: (checked: boolean) => void
+}
+
+interface ICheckboxMethods {
+  onHandlePress: () => void
+}
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
```
Comment (honghoangsts) [inline src/components/Checkbox/Checkbox.tsx:104]: remove `eslint-disable`

File: src/components/Checkbox/Checkbox.tsx [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+    useImperativeHandle(forwardedRef, () => ({onHandlePress}), [onHandlePress])
+
+    return (
+      <Container
+        testID="container"
+        {...rest}
+        style={style}
+        onPressIn={!disable ? bounceInEffect : null}
+        onPressOut={!disable ? bounceOutEffect : null}
+        onPress={!disable ? onHandlePress : null}>
```
Comment (honghoangsts) [inline src/components/Checkbox/Checkbox.tsx:231]: Why we don't pass direct `disabled` to component?

File: src/components/Checkbox/Checkbox.tsx [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+  /** the text of the checkbox */
+  text?: string
+  /** color when checkbox is checked, default #ffc484 */
+  fillColor?: string
+  /** define the status of checkbox */
+  isChecked?: boolean
+  /** color when checkbox is unchecked, default transparent */
+  unfillColor?: string
+  /** disable checkbox */
+  disable?: boolean
```
Comment (honghoangsts) [inline src/components/Checkbox/Checkbox.tsx:53]: use `disabled` from `BaseTouchableProps`

File: src/components/Checkbox/Checkbox.tsx [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        (!disableText || checkDisableTextType) &&
+        (textComponent || (
+          <TextContainer style={textContainerStyle} disable={disable} disableOpacity={disableOpacity}>
+            <Text testID="text" style={textStyle}>
+              {text}
+            </Text>
+          </TextContainer>
+        ))
+      )
+    }
```
Comment (honghoangsts) [inline src/components/Checkbox/Checkbox.tsx:212]: Can we simplify this logic?

File: src/components/Checkbox/Checkbox.tsx [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+        {renderCheckIcon()}
+        {renderCheckboxText()}
+      </Container>
+    )
+  },
+)
+
+Checkbox.displayName = 'Checkbox'
+
+export default memo(Checkbox)
```
Comment (honghoangsts) [inline src/components/Checkbox/Checkbox.tsx:241]: Why we need `memo`?

File: src/index.tsx [hunk 9]
```diff
@@ -20,6 +21,7 @@ export const RnBaseComponentView =
       }
 
 export const RnBaseButton = Button
+export const RnBaseCheckbox = Checkbox
```
Comment (honghoangsts) [inline src/index.tsx:24]: We don't need prefix `RnBase`

**Review by honghoangsts — CHANGES_REQUESTED** [review]
- Please move the magic number and magic string
- Use theme for default value of metrics and color.

File: src/components/Checkbox/Checkbox.tsx (hunk 2 above)
Comment (ngochuyduong) [reply src/components/Checkbox/Checkbox.tsx]: updated

**Comment by ngochuyduong** [comment]
> * Please move the magic number and magic string
> * Use theme for default value of metrics and color.

updated

File: src/components/Checkbox/Checkbox.tsx (hunk 7 above)
Comment (ngochuyduong) [reply src/components/Checkbox/Checkbox.tsx]: updated

**Review by loc-nguyenthien — APPROVED** [review]

File: src/components/Checkbox/Checkbox.tsx (hunk 5 above)
Comment (ngochuyduong) [reply src/components/Checkbox/Checkbox.tsx]: updated

File: src/components/Checkbox/Checkbox.tsx (hunk 4 above)
Comment (ngochuyduong) [reply src/components/Checkbox/Checkbox.tsx]: updated

File: example/ios/Podfile.lock (hunk 3 above)
Comment (ngochuyduong) [reply example/ios/Podfile.lock]: updated

File: src/components/Checkbox/Checkbox.tsx (hunk 6 above)
Comment (ngochuyduong) [reply src/components/Checkbox/Checkbox.tsx]: updated

File: src/components/Checkbox/Checkbox.tsx (hunk 8 above)
Comment (ngochuyduong) [reply src/components/Checkbox/Checkbox.tsx]: updated

File: src/index.tsx (hunk 9 above)
Comment (ngochuyduong) [reply src/index.tsx]: updated

**Review by anhle10051996 — APPROVED** [review]

**Comment by anhle10051996** [comment]
could you provide your test case result ?

**Comment by ngochuyduong** [comment]
> could you provide your test case result ?

Done

File: src/components/Checkbox/Checkbox.tsx [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+  marginLeft: props.theme?.sizes?.small,
+  opacity: props.disabled ? props.disableOpacity : 1,
+}))
+
+const IconContainer = styled(Animated.View)((props: IconContainerStyle) => ({
+  alignItems: 'center',
+  justifyContent: 'center',
+  width: props.size || props.theme?.sizes?.huge,
+  height: props.size || props.theme?.sizes?.huge,
+  borderRadius: (props.size || props.theme?.sizes?.huge) ?? 24 / 4,
```
Comment (honghoangsts) [inline src/components/Checkbox/Checkbox.tsx:222]: We shouldn't use magic number

File: src/components/Checkbox/Checkbox.tsx [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+      size,
+      iconStyle,
+      iconComponent,
+      iconImageStyle,
+      fillColor = theme.colors.primary,
+      unfillColor = theme.colors.transparent,
+      disableBuiltInState = false,
+      isChecked,
+      innerIconStyle,
+      checkIconImageSource = require('../../assets/images/check.png'),
```
Comment (honghoangsts) [inline src/components/Checkbox/Checkbox.tsx:109]: Please create constant for images

**Review by honghoangsts — CHANGES_REQUESTED** [review]

File: src/components/Checkbox/Checkbox.tsx (hunk 10 above)
Comment (ngochuyduong) [reply src/components/Checkbox/Checkbox.tsx:222]: updated

File: src/components/Checkbox/Checkbox.tsx (hunk 11 above)
Comment (ngochuyduong) [reply src/components/Checkbox/Checkbox.tsx]: updated

File: src/theme/base/colors.ts [hunk 12]
```diff
@@ -36,6 +36,7 @@ const colors = {
   coolGray: '#374151',
   blueGray: '#334155',
   dark: '#18181b',
+  transparent: '#00000000',
```
Comment (honghoangsts) [inline src/theme/base/colors.ts:39]: Why we don't use `'transparent'`

Comment (ngochuyduong) [reply src/theme/base/colors.ts:39]: Just want the hex code for consistency

**Comment by honghoangsts** [comment]
@ngochuyduong should we write test case for lines 132 and 136?

**Comment by ngochuyduong** [comment]
> @ngochuyduong should we write test case for lines 132 and 136?

update test cases

**Review by baonguyenhsts — APPROVED** [review]

**Review by vunguyendhSTS — CHANGES_REQUESTED** [review]

File: src/components/Checkbox/Checkbox.tsx [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+
+    const onHandlePress = useCallback(() => {
+      if (!disableBuiltInState) {
+        setChecked(prev => !prev)
+      }
+      syntheticBounceEffect()
+      onChange?.(!checked)
+    }, [disableBuiltInState, checked, onChange, syntheticBounceEffect])
+
+    useImperativeHandle(forwardedRef, () => ({onHandlePress}), [onHandlePress])
```
[the remark this answers is not in the document]
Comment (ThinhKimVo) [reply src/components/Checkbox/Checkbox.tsx:186]: @ngochuyduong Could you answer?

[the remark this answers is not in the document]
Comment (ThinhKimVo) [reply src/components/Checkbox/Checkbox.tsx:186]: I think we need a `value` prop.

[the remark this answers is not in the document]
Comment (ngochuyduong) [reply src/components/Checkbox/Checkbox.tsx:186]: we have the isChecked property to control the state of the component and the onChange callback when the checkbox changes state. we can based on the onChange callback. @vunguyendhSTS

[the remark this answers is not in the document]
Comment (ngochuyduong) [reply src/components/Checkbox/Checkbox.tsx:186]: btw, the isChecked is the same as the value prop that @ThinhKimVo mentioned

**Review by vunguyendhSTS — APPROVED** [review]

**Review by honghoangsts — APPROVED** [review]
